src/main/java/org/opensearch/ad/feature/SearchFeatureDao.java

```diff
@@ -190,67 +198,200 @@ public void getHighestCountEntities(AnomalyDetector detector, long startTime, lo
-            .aggregation(termsAgg)
+            .aggregation(bucketAggs)
             .trackTotalHits(false)
             .size(0);
         SearchRequest searchRequest = new SearchRequest().indices(detector.getIndices().toArray(new String[0])).source(searchSourceBuilder);
-        ActionListener<SearchResponse> termsListener = ActionListener.wrap(response -> {
-            Aggregations aggs = response.getAggregations();
-            if (aggs == null) {
-                listener.onResponse(Collections.emptyList());
-                return;
-            }
-            List<Entity> results = aggs
-                .asList()
-                .stream()
-                .filter(agg -> AGG_NAME_TERM.equals(agg.getName()))
-                .flatMap(agg -> ((Terms) agg).getBuckets().stream())
-                .map(bucket -> bucket.getKeyAsString())
-                .collect(Collectors.toList())
-                .stream()
-                .map(entityValue -> parseCategoricalField(entityValue, detector))
-                .collect(Collectors.toList());
-            listener.onResponse(results);
-        }, listener::onFailure);
-        client.search(searchRequest, termsListener);
+        client.search(searchRequest, new TopEntitiesListener(listener, detector, searchSourceBuilder));
     }
 
-    /**
-     * Precondition:
-     * {@code detector != null && detector.getCategoryField().size() > 0 }
-     *
-     * @param entityValue the representation of the entity's attributes. For
-     *  single-attribute entity, it is the value of the attribute like "server_1".
-     *  For multi-attribute entity, it is a map of attribute names and values like
-     *  "<code> {service=app_0, host=server_1}" </code>.
-     * @param detector Anomaly detector
-     * @return an Entity object corresponding to the entity
-     */
-    private Entity parseCategoricalField(String entityValue, AnomalyDetector detector) {
-        List<String> categoricalFields = detector.getCategoryField();
-        if (categoricalFields.size() == 1) {
-            return Entity.createSingleAttributeEntity(detector.getDetectorId(), detector.getCategoryField().get(0), entityValue);
+    class TopEntitiesListener implements ActionListener<SearchResponse> {
+        private ActionListener<List<Entity>> listener;
+        private AnomalyDetector detector;
+        private List<Entity> topEntities;
+        private SearchSourceBuilder searchSourceBuilder;
+
+        TopEntitiesListener(ActionListener<List<Entity>> listener, AnomalyDetector detector, SearchSourceBuilder searchSourceBuilder) {
+            this.listener = listener;
+            this.detector = detector;
+            this.topEntities = new ArrayList<>();
+            this.searchSourceBuilder = searchSourceBuilder;
+        }
+
+        @Override
+        public void onResponse(SearchResponse response) {
+            try {
+                Aggregations aggs = response.getAggregations();
+                if (aggs == null) {
```

Comment:
Is this okay to be null?

Reply:
It is not ok.  So I returned immediately without proceeding forward.

Reply:
Right, but then why isnt onFailure called?

Reply:
This would indicate some bug or some opensearch core changes that we are not aware of (we don't keep up-to-date with the large amounts of changes there). For example, they may change to if there are results return it; otherwise return null instead of an empty Aggregations as they currently do.  I am ok with fail fast or recover from the failure, log it, and return whatever we have.  No strong preference.  What about you?